### PR TITLE
util: enable golang profiling

### LINKED ISSUE #1699: Do golang profiling to find out any memory leak in cephcsi
Golang supports [profiling](https://blog.golang.org/pprof) , we can also make use of profiling and check are we leaking any memory and if required we can also do some optimization to improve the performance.

### PR DESCRIPTION
**Pull request by yati1998**
Add support for golang profiling.
Standard tools like go tool pprof and curl
work. example:
$ go tool pprof http://localhost:8080/debug/pprof/profile
$ go tool pprof http://localhost:8080/debug/pprof/heap
$ curl http://localhost:8080/debug/pprof/heap?debug=1

https://golang.org/pkg/net/http/pprof/ contains
more details about the pprof interface.

closes: #1699

Signed-off-by: Yati Padia <ypadia@redhat.com>

### CODE REVIEW

File: internal/util/httpserver.go [hunk 1]
```diff
@@ -24,3 +26,20 @@ func StartMetricsServer(c *Config) {
...
 	}
 }
+
+// EnableProfiling enables golang profiling.
+func EnableProfiling() {
+	for _, profile := range runtime_pprof.Profiles() {
+		name := profile.Name()
+		handler := pprof.Handler(name)
+
+		http.Handle(name, handler)
```
Comment (nixpanic) [inline internal/util/httpserver.go]: It would help if you can list the name/path where each handler can be reached. Anyone that has the profiling enabled will then be able to see what this particular executable supports (and new Golang versions might add/drop pprof profiles).

File: cmd/cephcsi.go [hunk 2]
```diff
@@ -90,6 +90,7 @@ func init() {
 		"skip image flattening if kernel support mapping of rbd images which has the deep-flatten feature")
 
 	flag.BoolVar(&conf.Version, "version", false, "Print cephcsi version information")
+	flag.BoolVar(&conf.EnableProfiling, "enableprofiling", true, "enable go profiling")
```
Comment (nixpanic) [inline cmd/cephcsi.go]: I am not sure we want it enabled by default. I think it is fine, but you probably should add a commit that adds the option for the `deploy/` and `charts/` yaml files.

File: cmd/cephcsi.go [hunk 3]
```diff
@@ -78,7 +78,7 @@ func init() {
 	flag.DurationVar(&conf.PollTime, "polltime", time.Second*pollTime, "time interval in seconds between each poll")
 	flag.DurationVar(&conf.PoolTimeout, "timeout", time.Second*probeTimeout, "probe timeout in seconds")
 
-	flag.BoolVar(&conf.EnableGRPCMetrics, "enablegrpcmetrics", false, "[DEPRECATED] enable grpc metrics")
+	flag.BoolVar(&conf.EnableGRPCMetrics, "enablegrpcmetrics", true, "[DEPRECATED] enable grpc metrics")
```
Comment (nixpanic) [inline cmd/cephcsi.go]: I understand that this is used for starting the http service. But it probably should be a different commit (and mention that the grpc metrics will be removed).

**Review by nixpanic — COMMENTED** [review]
Looks good in general. Some documentation would be good. Also a mentioning of the option in the deployment yaml files is missing.

File: internal/util/httpserver.go (hunk 1 above)
Comment (yati1998) [reply internal/util/httpserver.go]: okay

File: cmd/cephcsi.go (hunk 3 above)
Comment (yati1998) [reply cmd/cephcsi.go]: sure

**Review by nixpanic — DISMISSED** [review]
/retest ci/centos/mini-e2e-helm

File: cmd/cephcsi.go (hunk 3 above)
Comment (humblec) [reply cmd/cephcsi.go]: @yati1998 why we are flipping it back ?

File: cmd/cephcsi.go (hunk 2 above)
Comment (humblec) [reply cmd/cephcsi.go]: I dont think enabling `Profiling` by default is a good idea and also invite some security issues. Especially when we expose `/debug` in a cluster which can be used by any one in the cluster or outside depends on the production cluster setup.

**Review by humblec — DISMISSED** [review]

File: cmd/cephcsi.go (hunk 2 above)
Comment (yati1998) [reply cmd/cephcsi.go]: yes, I am working on that to add enable options to the yaml files.

File: cmd/cephcsi.go (hunk 3 above)
Comment (yati1998) [reply cmd/cephcsi.go]: For profiling we need server to get started and as Niels suggested, I have enabled it in different commit. Also, I will be working on removing the GRPC metrics later.

Comment (Madhu-1) [reply cmd/cephcsi.go]: disable by default.

File: cmd/cephcsi.go (hunk 2 above)
Comment (Madhu-1) [reply cmd/cephcsi.go]: disable it by default in code. if required enable it in yaml files

Comment (Madhu-1) [reply cmd/cephcsi.go]: add a check to start the HTTP server if the `enablegrpcmetrics` is disabled and profiling is enabled

File: internal/cephfs/driver.go [hunk 4]
```diff
@@ -164,5 +164,12 @@ func (fs *Driver) Run(conf *util.Config) {
 		util.WarningLogMsg("EnableGRPCMetrics is deprecated")
 		go util.StartMetricsServer(conf)
 	}
+	if conf.EnableProfiling {
+		if !conf.EnableGRPCMetrics {
+			go util.StartMetricsServer(conf)
+		}
+		util.WarningLogMsg("Registering profiling handler")
```
Comment (Madhu-1) [inline internal/cephfs/driver.go]: it should be the info log not warning log?

File: internal/util/httpserver.go [hunk 5]
```diff
@@ -24,3 +28,22 @@ func StartMetricsServer(c *Config) {
 		FatalLogMsg("failed to listen on address %v: %s", addr, err)
 	}
 }
+func addPath(name string, handler http.Handler) {
+	http.Handle(name, handler)
+	fmt.Fprintf(os.Stderr, "DEBUG: registered profiling handler on /debug/pprof/%s\n", name)
```
Comment (Madhu-1) [inline internal/util/httpserver.go]: use util.Info logger here.

File: cmd/cephcsi.go [hunk 6]
```diff
@@ -90,6 +90,7 @@ func init() {
 		"skip image flattening if kernel support mapping of rbd images which has the deep-flatten feature")
 
 	flag.BoolVar(&conf.Version, "version", false, "Print cephcsi version information")
+	flag.BoolVar(&conf.EnableProfiling, "enableprofiling", false, "enable go profiling")
```
Comment (Madhu-1) [inline cmd/cephcsi.go:93]: helm template update is missing for this one.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
Enable this flag in E2E and make sure you are getting the profiling response.

File: internal/cephfs/driver.go (hunk 4 above)
Comment (yati1998) [reply internal/cephfs/driver.go]: yes. will update this.

File: internal/util/httpserver.go (hunk 5 above)
Comment (yati1998) [reply internal/util/httpserver.go]: ok.

File: cmd/cephcsi.go (hunk 6 above)
Comment (yati1998) [reply cmd/cephcsi.go:93]: yes, working on helm charts. Will update the PR once that is done. and Also, will be adding documentation once done.

File: internal/util/httpserver.go [hunk 7]
```diff
@@ -24,3 +26,22 @@ func StartMetricsServer(c *Config) {
 		FatalLogMsg("failed to listen on address %v: %s", addr, err)
 	}
 }
+func addPath(name string, handler http.Handler) {
+	http.Handle(name, handler)
+	DebugLogMsg("DEBUG: registered profiling handler on /debug/pprof/%s\n", name)
```
Comment (nixpanic) [inline internal/util/httpserver.go:31]: nit: drop `DEBUG:` from this message, it will be tagged with debug level already too through `DebugLogMsg`

**Review by nixpanic — DISMISSED** [review]
LGTM, you may want to update the debug message a little though.

The **ci/centos/k8s-e2e-external-storage** CI job is optional, and does not need to pass. #2036 has been posted to address its failure.

File: internal/util/httpserver.go (hunk 7 above)
Comment (yati1998) [reply internal/util/httpserver.go:31]: okay sure.

File: cmd/cephcsi.go (hunk 3 above)
Comment (nixpanic) [reply cmd/cephcsi.go]: it is disabled by default in the current state

File: cmd/cephcsi.go (hunk 2 above)
Comment (nixpanic) [reply cmd/cephcsi.go]: This has been done already.

**Review by Madhu-1 — DISMISSED** [review]

**Review by nixpanic — DISMISSED** [review]

**Review by Madhu-1 — APPROVED** [review]

File: charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml [hunk 8]
```diff
@@ -75,6 +75,9 @@ spec:
             - "--drivername=$(DRIVER_NAME)"
 {{- if .Values.topology.enabled }}
             - "--domainlabels={{ .Values.topology.domainLabels | join "," }}"
+{{- end }}
+{{- if .Values.nodeplugin.enableprofiling.enabled }}
+            - "--enableprofiling={{ .Values.nodeplugin.enableprofiling.enabled }}"
```
Comment (Madhu-1) [inline charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml]:
```suggestion
- "--enableprofiling={{ .Values.nodeplugin.profiling.enabled }}"
```

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
a small suggestion to the helm charts for better readability.

File: charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml (hunk 8 above)
Comment (Madhu-1) [reply charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml]: Please take care of this in other places also for the helm. and also Enable this in helm testing lets see if nothing breaks.

Comment (yati1998) [reply charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml]: sure. Will update it.

**Review by nixpanic — DISMISSED** [review]

**Comment by Rakshith-R** [comment]
@yati1998 please rebase pr on top of devel for ci fix.

**Review by Madhu-1 — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]
